File: src/signals/protocol.ts
/**
 * DeFiLlama protocol revenue fetcher.
 *
 * Endpoint: https://api.llama.fi/overview/fees
 * Returns daily fee revenue for 2000+ DeFi protocols.
 *
 * Signal logic:
 *   - Today's fees / 7-day average fee > threshold → revenue spike
 *   - Token price hasn't moved yet (price change < 12%) → mispricing window
 *   - TVL is stable or growing → protocol retaining users
 *
 * This is a fundamentals-driven signal — holds hours to days, not minutes.
 * Maps to the protocol_revenue strategy in strategies/protocol-revenue.ts.
 */

import { log } from '../storage/database.js';

// ─── Types ────────────────────────────────────────────────────────────────────

interface LlamaProtocol {
  name: string;
  displayName?: string;
  disabled?: boolean;
  total24h: number | null;
  total7d: number | null;
  totalAllTime: number | null;
  chains: string[];
  gecko_id?: string;
  symbol?: string;
}

interface LlamaFeesResponse {
  protocols: LlamaProtocol[];
}

export interface ProtocolRevenueData {
  protocol: string;
  symbol: string;
  revenue24h: number;
  revenue7dAvg: number;
  revenueMultiple: number; // 24h / 7dAvg
  sampledAt: number;
}

// ─── Symbol overrides for tokens not in DeFiLlama gecko_id ───────────────────

const PROTOCOL_SYMBOL_MAP: Record<string, string> = {
  uniswap:   'UNI',
  aave:      'AAVE',
  curve:     'CRV',
  compound:  'COMP',
  lido:      'LDO',
  makerdao:  'MKR',
  synthetix: 'SNX',
  gmx:       'GMX',
  dydx:      'DYDX',
  ondo:      'ONDO',
  maple:     'MPL',
};

function resolveSymbol(p: LlamaProtocol): string | undefined {
  const name = p.name.toLowerCase().replace(/[^a-z0-9]/g, '');
  if (PROTOCOL_SYMBOL_MAP[name]) return PROTOCOL_SYMBOL_MAP[name];
  if (p.symbol && p.symbol.length > 0) return p.symbol.toUpperCase();
  return undefined;
}

// ─── Fetch ────────────────────────────────────────────────────────────────────

let cached: ProtocolRevenueData[] = [];
let lastFetchAt = 0;
const CACHE_TTL_MS = 3_600_000; // 1 hour — data updates daily

export async function fetchProtocolRevenue(): Promise<ProtocolRevenueData[]> {
  if (Date.now() - lastFetchAt < CACHE_TTL_MS) return cached;

  try {
    const res = await fetch('https://api.llama.fi/overview/fees?excludeTotalDataChartBreakdown=true', {
      signal: AbortSignal.timeout(15_000),
    });
    if (!res.ok) {
      log('warn', `DeFiLlama fees fetch failed: ${res.status}`);
      return cached;
    }

    const data = await res.json() as LlamaFeesResponse;
    const results: ProtocolRevenueData[] = [];

    for (const protocol of data.protocols) {
      if (protocol.disabled) continue;
      if (!protocol.total24h || !protocol.total7d) continue;
      if (protocol.total24h < 1_000) continue; // filter micro-protocols

      const symbol = resolveSymbol(protocol);
      if (!symbol) continue;

      const revenue7dAvg = protocol.total7d / 7;
      const revenueMultiple = revenue7dAvg > 0 ? protocol.total24h / revenue7dAvg : 0;

      results.push({
        protocol: protocol.displayName ?? protocol.name,
        symbol,
        revenue24h: protocol.total24h,
        revenue7dAvg,
        revenueMultiple,
        sampledAt: Date.now(),
      });
    }

    // Sort by revenue multiple descending — highest spike at top
    results.sort((a, b) => b.revenueMultiple - a.revenueMultiple);

    lastFetchAt = Date.now();
    cached = results;
    log('info', `DeFiLlama: loaded ${results.length} protocol revenue records`);
    return results;
  } catch (err) {
    log('warn', `DeFiLlama network error: ${String(err)}`);
    return cached;
  }
}

File: src/signals/protocol.py
"""DeFiLlama protocol revenue fetcher."""

import re
import time
from dataclasses import dataclass

import requests

from src.signals._circuit_breaker import CircuitBreaker
from src.storage.database import log


@dataclass
class ProtocolRevenueData:
    protocol: str
    symbol: str
    revenue_24h: float
    revenue_7d_avg: float
    revenue_multiple: float
    sampled_at: float


PROTOCOL_SYMBOL_MAP: dict[str, str] = {
    "uniswap": "UNI", "aave": "AAVE", "curve": "CRV", "compound": "COMP",
    "lido": "LDO", "makerdao": "MKR", "synthetix": "SNX", "gmx": "GMX",
    "dydx": "DYDX", "ondo": "ONDO", "maple": "MPL",
}

_cached: list[ProtocolRevenueData] = []
_last_fetch_at: float = 0
_CACHE_TTL_MS = 3_600_000
_breaker = CircuitBreaker("protocol")


def _resolve_symbol(protocol: dict) -> str | None:
    name = re.sub(r"[^a-z0-9]", "", protocol.get("name", "").lower())
    if name in PROTOCOL_SYMBOL_MAP:
        return PROTOCOL_SYMBOL_MAP[name]
    sym = protocol.get("symbol")
    if sym:
        return sym.upper()
    return None


def fetch_protocol_revenue() -> list[ProtocolRevenueData]:
    global _cached, _last_fetch_at
    now = time.time() * 1000
    if now - _last_fetch_at < _CACHE_TTL_MS:
        return _cached

    # Staleness warning
    if _cached and _last_fetch_at > 0 and now - _last_fetch_at > 2 * _CACHE_TTL_MS:
        log("warn", f"Protocol revenue data is stale (last fetch {(now - _last_fetch_at) / 60_000:.0f}m ago)")

    if not _breaker.can_call():
        log("warn", "Protocol revenue circuit breaker OPEN — returning cached data")
        return _cached

    try:
        res = requests.get(
            "https://api.llama.fi/overview/fees?excludeTotalDataChartBreakdown=true",
            timeout=15,
        )
        if res.status_code != 200:
            log("warn", f"DeFiLlama fees fetch failed: {res.status_code}")
            _breaker.record_failure()
            return _cached

        data = res.json()
        results: list[ProtocolRevenueData] = []

        for protocol in data.get("protocols", []):
            if protocol.get("disabled"):
                continue
            total_24h = protocol.get("total24h")
            total_7d = protocol.get("total7d")
            if not total_24h or not total_7d:
                continue
            if total_24h < 1000:
                continue

            symbol = _resolve_symbol(protocol)
            if not symbol:
                continue

            revenue_7d_avg = total_7d / 7
            revenue_multiple = total_24h / revenue_7d_avg if revenue_7d_avg > 0 else 0

            results.append(ProtocolRevenueData(
                protocol=protocol.get("displayName") or protocol["name"],
                symbol=symbol,
                revenue_24h=total_24h,
                revenue_7d_avg=revenue_7d_avg,
                revenue_multiple=revenue_multiple,
                sampled_at=now,
            ))

        results.sort(key=lambda x: x.revenue_multiple, reverse=True)
        _last_fetch_at = now
        _cached = results
        _breaker.record_success()
        log("info", f"DeFiLlama: loaded {len(results)} protocol revenue records")
        return results

    except Exception as err:
        log("warn", f"DeFiLlama network error: {err}")
        _breaker.record_failure()
        return _cached
